File: src/BreakfastProvider.Api/Reporting/CustomerFeedbackAlertService.cs
using BreakfastProvider.Api.HttpClients;
using BreakfastProvider.Api.Services;
using BreakfastProvider.Api.Telemetry;
using MongoDB.Bson.Serialization.Attributes;
using MongoDB.Driver;

namespace BreakfastProvider.Api.Reporting;

public interface ICustomerFeedbackAlertService
{
    Task ProcessFeedbackAsync(Guid feedbackId, string customerName, string recipeName, int rating, string comments, DateTime receivedAt, CancellationToken cancellationToken = default);
}

public class CustomerFeedbackAlertService(
    IMongoClient mongoClient,
    INotificationClient notificationClient,
    IHttpClientFactory httpClientFactory,
    ILogger<CustomerFeedbackAlertService> logger) : ICustomerFeedbackAlertService
{
    private IMongoCollection<CustomerFeedbackAlertDocument> Collection =>
        mongoClient.GetDatabase("BreakfastDb").GetCollection<CustomerFeedbackAlertDocument>("feedback_alerts");

    public async Task ProcessFeedbackAsync(Guid feedbackId, string customerName, string recipeName, int rating, string comments, DateTime receivedAt, CancellationToken cancellationToken = default)
    {
        using var activity = DiagnosticsConfig.ActivitySource.StartActivity("CustomerFeedbackAlertService.ProcessFeedback");

        // Step 1: Write to MongoDB
        var doc = new CustomerFeedbackAlertDocument
        {
            AlertId = feedbackId.ToString(),
            CustomerName = customerName,
            RecipeName = recipeName,
            Rating = rating,
            Comments = comments,
            ReceivedAt = receivedAt,
            ProcessedAt = DateTime.UtcNow
        };

        await Collection.InsertOneAsync(doc, cancellationToken: cancellationToken);
        logger.LogInformation("Feedback alert {FeedbackId} stored in MongoDB for recipe {RecipeName}", feedbackId, recipeName);

        // Step 2: Send notification via gRPC
        var (success, notificationId) = await notificationClient.SendOrderConfirmationAsync(
            feedbackId.ToString(), customerName, rating, cancellationToken);

        logger.LogInformation("Notification sent for feedback {FeedbackId}: Success={Success}, NotificationId={NotificationId}",
            feedbackId, success, notificationId);

        // Step 3: Call Supplier Service to log the feedback
        var supplierClient = httpClientFactory.CreateClient(HttpClientNames.SupplierService);
        await supplierClient.PostAsJsonAsync("ingredients/feedback", new
        {
            FeedbackId = feedbackId,
            RecipeName = recipeName,
            Rating = rating,
            CustomerName = customerName
        }, cancellationToken);

        logger.LogInformation("Supplier service notified about feedback {FeedbackId} for recipe {RecipeName}", feedbackId, recipeName);
    }
}

public class CustomerFeedbackAlertDocument
{
    [BsonId]
    public string AlertId { get; set; } = string.Empty;
    public string CustomerName { get; set; } = string.Empty;
    public string RecipeName { get; set; } = string.Empty;
    public int Rating { get; set; }
    public string Comments { get; set; } = string.Empty;
    public DateTime ReceivedAt { get; set; }
    public DateTime ProcessedAt { get; set; }
}

File: java/breakfast-provider-app/src/main/java/io/lemonlion/breakfast/reporting/CustomerFeedbackAlertService.java
package io.lemonlion.breakfast.reporting;

import com.mongodb.client.MongoClient;
import io.lemonlion.breakfast.config.DownstreamConfig;
import io.lemonlion.breakfast.model.event.CustomerFeedbackReceivedEvent;
import io.lemonlion.breakfast.notification.NotificationClient;
import java.time.Instant;
import java.util.Date;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.UUID;
import org.bson.Document;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.web.client.RestTemplateBuilder;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

/**
 * Twin of C# {@code CustomerFeedbackAlertService}: on a consumed feedback event, store it in MongoDB
 * ({@code feedback_alerts}), send a gRPC notification, and POST it to the Supplier service.
 */
@Service
public class CustomerFeedbackAlertService {

    private static final Logger log = LoggerFactory.getLogger(CustomerFeedbackAlertService.class);

    private final MongoClient mongoClient;
    private final NotificationClient notificationClient;
    private final RestTemplate restTemplate;
    private final String supplierUrl;

    public CustomerFeedbackAlertService(MongoClient mongoClient, NotificationClient notificationClient,
                                        RestTemplateBuilder builder, DownstreamConfig downstreamConfig) {
        this.mongoClient = mongoClient;
        this.notificationClient = notificationClient;
        this.restTemplate = builder.build();
        this.supplierUrl = downstreamConfig.getSupplierServiceUrl();
    }

    public void processFeedback(CustomerFeedbackReceivedEvent feedback) {
        Document doc = new Document("_id", feedback.feedbackId().toString())
                .append("customerName", feedback.customerName())
                .append("recipeName", feedback.recipeName())
                .append("rating", feedback.rating())
                .append("comments", feedback.comments())
                .append("receivedAt", feedback.receivedAt() == null ? null : Date.from(feedback.receivedAt()))
                .append("processedAt", Date.from(Instant.now()));
        mongoClient.getDatabase("BreakfastDb").getCollection("feedback_alerts").insertOne(doc);

        notificationClient.sendOrderConfirmation(feedback.feedbackId(), feedback.customerName(), feedback.rating());

        Map<String, Object> body = new LinkedHashMap<>();
        body.put("feedbackId", feedback.feedbackId());
        body.put("recipeName", feedback.recipeName());
        body.put("rating", feedback.rating());
        body.put("customerName", feedback.customerName());
        restTemplate.postForEntity(supplierUrl + "/ingredients/feedback", body, Void.class);

        log.info("Processed customer feedback {} for recipe {}", feedback.feedbackId(), feedback.recipeName());
    }
}
